src/cryptonote_core/service_node_list.h

```diff
@@ -104,6 +104,8 @@ namespace service_nodes
     cryptonote::account_public_address operator_address;
     uint32_t                           public_ip;
     uint16_t                           storage_port;
+    uint64_t                           last_ip_change_height; // The height of the last quorum penalty for changing IPs
+    std::vector<std::pair<uint32_t, uint64_t>> proof_public_ips; // (not serialized)
```

Comment:
Making this unbounded makes me a bit uneasy. A std::array of size 8 would be sufficient. Any more is clearly someone trying to abuse the system.

Reply:
Actually it can probably just be a std::array of size 2 that just keeps the most recent 2 IPs seen (and updates the timestamp when it sees one again).